### PR TITLE
Support TO EVALUATE execution for Tensorflow and Keras

### PR DESCRIPTION
**Pull request by typhoonzero**
TODO:
1. support xgboost
1. support running on PAI.

### CODE REVIEW

File: pkg/ir/ir.go [hunk 1]
```diff
@@ -156,6 +157,30 @@ func (cl *ExplainStmt) IsExtended() bool { return true }
 // GetOriginalSQL returns the original SQL statement used to get current IR result
 func (cl *ExplainStmt) GetOriginalSQL() string { return cl.OriginalSQL }
 
+// EvaluateStmt is the intermediate representation for code generation of a analysis job
```
Comment (lhw362950217) [inline pkg/ir/ir.go]: of a analysis job -> of an evaluation job

File: pkg/ir/ir.go [hunk 2]
```diff
@@ -156,6 +157,30 @@ func (cl *ExplainStmt) IsExtended() bool { return true }
...
+	Select           string
+	Attributes       map[string]interface{}
+	ModelName        string
+	Label            FeatureColumn
+	Into             string
+	TmpEvaluateTable string
+	TrainStmt        *TrainStmt
+}
+
+// Execute generates and executes code for ExplainStmt
```
Comment (lhw362950217) [inline pkg/ir/ir.go]: for ExplainStmt -> for EvaluateStmt

File: pkg/sql/codegen/tensorflow/codegen.go [hunk 3]
```diff
@@ -521,6 +521,47 @@ func Explain(stmt *ir.ExplainStmt, session *pb.Session) (string, error) {
 	return program.String(), nil
 }
 
+// Evaluate generates a Python program to explain a trained model.
```
Comment (lhw362950217) [inline pkg/sql/codegen/tensorflow/codegen.go]: explain -> evaluate

File: pkg/sql/codegen/tensorflow/template_evaluate.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+package tensorflow
+
+import "sqlflow.org/sqlflow/pkg/ir"
+
+type evaluateFiller struct {
+	DataSource  string
+	Select      string
+	ResultTable string
+	// below members comes from trainStmt
```
Comment (weiguoz) [inline pkg/sql/codegen/tensorflow/template_evaluate.go]: typo: comes -> come

File: pkg/sql/codegen/tensorflow/template_evaluate.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+         save="{{.Save}}",
+         batch_size=1,
+         validation_steps=None,
+         verbose=0,
+         hdfs_namenode_addr="{{.HDFSNameNodeAddr}}",
+         hive_location="{{.HiveLocation}}",
+         hdfs_user="{{.HDFSUser}}",
+         hdfs_pass="{{.HDFSPass}}",
+         is_pai="{{.IsPAI}}" == "true",
+		 pai_table="{{.PAIEvaluateTable}}")
```
Comment (weiguoz) [inline pkg/sql/codegen/tensorflow/template_evaluate.go]: indent

File: pkg/sql/ir_generator.go [hunk 6]
```diff
@@ -339,6 +342,54 @@ func generateExplainStmt(slct *parser.SQLFlowSelectStmt, connStr, modelDir strin
...
+	var trainStmt *ir.TrainStmt
+	if getTrainStmtFromModel {
+		trainStmt, err = generateTrainStmtByModel(slct, connStr, cwd, modelDir, slct.ModelToEvaluate)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	resultTable := slct.EvaluateInto
+	if err != nil {
```
Comment (weiguoz) [inline pkg/sql/ir_generator.go]: What's the `err` about?

File: pkg/sql/submitter.go [hunk 7]
```diff
@@ -218,6 +218,71 @@ func (s *defaultSubmitter) ExecuteExplain(cl *ir.ExplainStmt) error {
 	return nil
 }
 
+func (s *defaultSubmitter) ExecuteEvaluate(cl *ir.EvaluateStmt) error {
+	// NOTE(typhoonzero): model is already loaded under s.Cwd
+	var code string
+	var err error
```
Comment (weiguoz) [inline pkg/sql/submitter.go]: I would suggest merging these 2 lines into L229: `code, err := tensorflow.Evaluate(cl, s.Session)`

File: pkg/sql/submitter.go [hunk 8]
```diff
@@ -218,6 +218,71 @@ func (s *defaultSubmitter) ExecuteExplain(cl *ir.ExplainStmt) error {
...
+
+	if isXGBoostModel(cl.TrainStmt.Estimator) {
+		return fmt.Errorf("XGBoost evaluation is not supported now, will be available soon")
+	}
+	code, err = tensorflow.Evaluate(cl, s.Session)
```
Comment (weiguoz) [inline pkg/sql/submitter.go]: Check if err?
It looks the Line L248's error checking is unnecessary.

File: python/sqlflow_submitter/tensorflow/evaluate.py [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+                                  db_generator, pai_maxcompute_db_generator,
+                                  parseMaxComputeDSN)
+
+from . import metrics
+from .get_tf_version import tf_is_version2
+from .input_fn import get_dataset_fn
+from .pai_distributed import define_tf_flags, set_oss_environs
+from .set_log_level import set_log_level
+from .train_estimator import estimator_train_and_save
+from .train_keras import keras_train_and_save
```
Comment (weiguoz) [inline python/sqlflow_submitter/tensorflow/evaluate.py]: It looks like several packages imported but not used, such as: `functools, glob, json... keras_train_and_save `

**Review by weiguoz — APPROVED** [review]
LGTM

**Pull request merged**